src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.Linux.cs

```diff
@@ -358,13 +412,40 @@ private void Initialize()
                     return;
                 }
 
-                int fileDescriptor = Interop.open(GpioMemoryFilePath, FileOpenFlags.O_RDWR | FileOpenFlags.O_SYNC);
+                // try and open /dev/gpiomem
+                fileDescriptor = Interop.open(GpioMemoryFilePath, FileOpenFlags.O_RDWR | FileOpenFlags.O_SYNC);
                 if (fileDescriptor == -1)
                 {
-                    throw new IOException($"Error {Marshal.GetLastWin32Error()} initializing the Gpio driver.");
-                }
+                    // if the failure is NOT because /dev/gpiomem doesn't exist then throw an exception at this point.
+                    // if it were anything else then it is probably best not to try and use /dev/mem on the basis that
+                    // it would be better to solve the issue rather than use a method that requires root privileges
+                    if(Marshal.GetLastWin32Error() != Linux_Enoent)
+                    {
+                        throw new IOException($"Error {Marshal.GetLastWin32Error()} initializing the Gpio driver.");
+                    }
+
+                    // if /dev/gpiomem doesn't seem to be available then let's try /dev/mem
+                    fileDescriptor = Interop.open(MemoryFilePath, FileOpenFlags.O_RDWR | FileOpenFlags.O_SYNC);
+                    if (fileDescriptor == -1)
+                    {
+                        throw new IOException($"Error {Marshal.GetLastWin32Error()} initializing the Gpio driver.");
+                    }
+                    else // success so set the offset into memory of the gpio registers
+                    {
+                        gpioRegisterOffset = GetPeripheralBaseAddress();
+
+                        if(gpioRegisterOffset == ~0U)
+                        {
+                            throw new Exception($"Error - Unable to determine peripheral base address.");
+                        }
+
+                        // add on the offset from the peripheral base address to point to the gpio registers
+                        gpioRegisterOffset += GpioPeripheralOffset;
+                    }
 
-                IntPtr mapPointer = Interop.mmap(IntPtr.Zero, Environment.SystemPageSize, (MemoryMappedProtections.PROT_READ | MemoryMappedProtections.PROT_WRITE), MemoryMappedFlags.MAP_SHARED, fileDescriptor, GpioRegisterOffset);
+                }
+                    
+                IntPtr mapPointer = Interop.mmap(IntPtr.Zero, Environment.SystemPageSize, (MemoryMappedProtections.PROT_READ | MemoryMappedProtections.PROT_WRITE), MemoryMappedFlags.MAP_SHARED, fileDescriptor, (int)gpioRegisterOffset);
```

Comment:
could this be using https://docs.microsoft.com/en-us/dotnet/api/system.io.memorymappedfiles.memorymappedfile?view=netcore-3.0 instead?

Reply:
@krwq  @joperezr 

Perhaps but I think that this would be out of scope for this PR as here we are trying to ensure that Gpio works with /dev/men as a fallback but to move away form interop for we would need to ask some wider questions about the merits of both methods.

Reply:
I'm fine with keeping this Interop for now, and then we can move to using memorymappedfiles in the future as a separate issue.

Reply:
please file an issue on this

Reply:
Issue #686 filed